src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -4,92 +4,91 @@
  */
 package com.vmware.vipclient.i18n.messages.service;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.ListIterator;
-import java.util.Map;
-
+import com.vmware.vipclient.i18n.VIPCfg;
+import com.vmware.vipclient.i18n.base.DataSourceEnum;
+import com.vmware.vipclient.i18n.messages.dto.BaseDTO;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.vmware.vipclient.i18n.VIPCfg;
-import com.vmware.vipclient.i18n.base.DataSourceEnum;
-import com.vmware.vipclient.i18n.base.cache.Cache;
-import com.vmware.vipclient.i18n.base.cache.FormatCacheItem;
-import com.vmware.vipclient.i18n.messages.api.opt.server.RemoteLocaleOpt;
-import com.vmware.vipclient.i18n.util.JSONUtils;
+import java.util.HashMap;
+import java.util.ListIterator;
+import java.util.Map;
 
 public class LocaleService {
 
     Logger                      logger        = LoggerFactory.getLogger(LocaleService.class.getName());
-    private static final String REGION_PREFIX = "region_";
     public static final String DISPN_PREFIX  = "dispn_";
+    private BaseDTO dto = null;
 
-    public LocaleService() {
+    public LocaleService(BaseDTO dto) {
+        this.dto = dto;
     }
 
-    public Map<String, Map<String, String>> getTerritoriesFromCLDR(
-            List<String> languages) {
-        
-        Map<String, Map<String, String>> respMap = new HashMap<String, Map<String, String>>();
-        for (String language : languages) {
-            language = language.toLowerCase();
-            Map<String, String> regionMap = null;
-            logger.trace("look for region list of '" + language + "' from cache");
-            Cache c = VIPCfg.getInstance().getCacheManager()
-                    .getCache(VIPCfg.CACHE_L2);
-            if (c != null) {
-            	FormatCacheItem cacheItem = (FormatCacheItem) c.get(REGION_PREFIX
-                        + language);    
-                regionMap = cacheItem == null ? regionMap : cacheItem.getCachedData();
-            }
-            if (regionMap != null) {
-                respMap.put(language, regionMap);
-                continue;
-            }
-            logger.trace("get region list of '" + language
-                    + "' data from backend");
-            Map<String, String> tmpMap = new RemoteLocaleOpt()
-				        .getTerritoriesFromCLDR(language);
-            regionMap = JSONUtils.map2SortMap(tmpMap);
-            respMap.put(language, regionMap);
-            if (c != null) {
-            	FormatCacheItem cacheItem = new FormatCacheItem(regionMap);
-                c.put(REGION_PREFIX + language, cacheItem);
-            }
+    public Map<String, String> getRegionsFromCLDR(String locale){
```

Comment:
- Remove locale parameter, make it a property of LocaleDTO.
As we discussed offline, add a new class LocaleDTO in case parameter increase, but won’t put locale into it.
Because 'locale' is a frequently changed parameter, put it into DTO will need new service instance multiple times when its value changed, and new instance is time consuming.

- Use fallbackLocalesIter as parameter and recursively call the method to apply locale fallback logic.
I can get the fallback iterator in service layer, no need to get it in message layer and pass it, xiaochao also suggest this.